feat: add doc for sas-auth-browser (#81)
sassoftware/visual-analytics-sdk: sas-browser-auth/README.md

# Overview

> The `sas-auth-browser` package provides an api that allows for single page applications to ensure that a user is authenticated before making api calls. Currently only cookie based authentication is supported.

Reviewer: Is calling out "single page applications" here necessary, or should we be more generic.  Also, this doc calls out "making api calls" but not the combined uses with VASDK or Content-SDK.  
Maybe something like
```suggestion
The `sas-auth-browser` package provides an api that allows for an application or web page to ensure that a user is authenticated before making SAS api calls or using SAS web components. Currently only cookie based authentication is supported.
```

> The `sas-auth-browser` package provides an api that allows for an application or web page to ensure that a user is authenticated before making SAS api calls or using SAS web components. Currently only cookie based authentication is supported.

Reviewer 2: A couple of minor changes: (1) Make sure API is in all caps in the first two sentences. (2) In the last sentence, add a hyphen between cookie and based. It should be cookie-based.

Reviewer: @ckedwards, I unresolved this one.  API needs to be capitalized everywhere.

Author: I did a find and replace for this, but I think I had a bad merge afterwards. Fixed.

# Prerequisites

> In order to successfully make browser based REST calls to viya endpoints using cookies, you will need to enable CORS, Cross-site cookies and CSRF web security settings on the server for your single page application. For more information please see the SAS® Visual Analytics SDK [Viya setup guide](https://developer.sas.com/sdk/va/docs/guides/viya-setup/).

Reviewer 2: A couple of minor changes: (1) Add a hyphen between single and page. It should be single-page application. (2) Drop "please" in the last sentence. It should be "For more information, see..."

> In order to successfully make browser based REST calls to SAS Viya endpoints using cookies, you will need to enable CORS, Cross-site cookies and CSRF web security settings on the server for your single page application. For more information please see the SAS® Visual Analytics SDK [SAS Viya setup guide](https://developer.sas.com/sdk/va/docs/guides/viya-setup/).

Reviewer: The "on the server for your single page application" confuses this a bit.  I think it is clearer without it.  @joylashley, any thoughts on this suggestion?
```suggestion
In order to successfully make browser based REST calls to SAS Viya endpoints using cookies, you will need to enable CORS, Cross-site cookies and CSRF web security settings. For more information please see the SAS® Visual Analytics SDK [SAS Viya setup guide](https://developer.sas.com/sdk/va/docs/guides/viya-setup/).
```

Reviewer 2: I like your suggestion of removing it, @brmorr.

# Installation

```
npm install @sassoftware/sas-auth-browser
```

# CDN

```html
<script async src="https://cdn.developer.sas.com/packages/sas-auth-browser/latest/dist/index.min.js"></script>
```

## Usage

> A simple example that uses sas-auth-browser in the [examples](./examples) directory.

Reviewer 3: This is redundant with the `Examples` section below. Was that intentional?

> Additionally, before a rest api is made to the viya server, you should first check to see if the cookie session is still valid.

Reviewer: Should it be "rest api call"?
```suggestion
Additionally, before a rest api call is made to the viya server, you should first check to see if the cookie session is still valid.
```

> Additionally, before a rest api call is made to the viya server, you should first check to see if the cookie session is still valid.

Reviewer 2: SAS Viya is a trademarked name, so SAS always has to be included. In this sentence, change "viya server" to SAS Viya server.

```ts
const sasAuthInstance = sasAuthBrowser.createCookieAuthenticationCredentialInstance({
  endpoint,
});
async function callViyaApi() {
  try {
    await sasAuthInstance.checkAuthenticated();
  } catch {

>     // Note: if the user closes the popup an uncaught exception will occur.

Reviewer 2: Minor change: Capitalize the first word in the note. "If the user..."

    await sasAuthInstance.loginPopup();
  }
  // Start making rest calls!
}
```

# API

> ## `createCookieAuthenticationCredentialInstance(configuration): CookieAuthenticationCredential`

Author: create* and `new` are two ways of making the same object, so I was thinking that wouldn't be too much of a leap. However, I ended up duplicating config because it was simple enough.

Creates a new CookieAuthenticationCredential instance. 

## CookieAuthenticationCredential

The CookieAuthenticationCredential class provides functions to check the authentication status of a given endpoint

> ### `new CookieAuthenticationCredential(configuration)`

Author: It is available for esm. So create vs new is dependent on esm vs global export.

* `configuration: object`

>   * `endpoint: string` The Viya server you are authenticating against.

Reviewer 2: Add SAS before "Viya server".

>   * `url: string` The url of SAS Viya server you are authenticating against.

Reviewer: Not sure how this is worded in other doc.  Are we authenticating "against" a server, or authenticating "with"? @joylashley ?

Reviewer 2: I think "with" is better than "against". But, see if this works:  "The URL of the SAS Viya server that is being authenticated."

Make sure all instances of "URL" are in all caps. I missed this in my earlier review.

Author: The server is not being authenticated, but rather the user is authenticating (by entering their user name and password etc). Can you provide any wording that reflects that it is a user that is authenticating?

Reviewer 2: Sorry I missed this earlier. If the user is authenticating, it's probably better to say, "The URL of the SAS Viya server that you are authenticating with."

>   * `guest: boolean` Automatically log in a guest if no user session is found.

Reviewer 2: Change to "Automatically log in as a guest, if no user session is found."

>   * `guest: boolean` Automatically log in a guest if no user session is found.

Reviewer: *as
```suggestion
  * `guest: boolean` Automatically log in as guest if no user session is found.
```

    * default: false 
#### Methods

##### `checkAuthenticated(): Promise<void>`

> If the server endpoint has been checked, use the cached value. Otherwise check the server endpoint to see if a user is logged in.

Reviewer 2: Add a comma after the word "Otherwise" in the second sentence.

> If the server endpoint has been checked, use the cached value. Otherwise check the server endpoint to see if a user is logged in.

Reviewer: That is a good point @cojhal.  This leaves me (and probably the developer) confused on when you should use `checkAuthenticated` vs `ensureAuthenticated`.

> If the server endpoint has been checked, use the cached value. Otherwise check the server endpoint to see if a user is logged in.

Author: i think i could just remove the cache discussion here and ensureAuthenticated. I added a `invalidateCache` (currently undocumented) which makes ensureAuthenticated unnecessary.

> Throws: if no user is authenticated.

Reviewer 3: When you view the rendered markdown, the `Throws` lines are concatted to the end of the previous line.

Author: `rejects` might be a better option.

Reviewer: It looks like one example was changed to `rejects`, but the one for checkAuthenticated still uses `throws`.  Wasn't sure if you meant to change both.

Author: Both should be rejects.

##### `ensureAuthenticated(): Promise<void>`

> Bypasses the cache and checked the endpoint to see if a user is authenticated.

Reviewer 2: Change "checked" to "checks".

> Throws: if no user is authenticated.

Reviewer 2: @brmorr, is it consistent to not have a an initial cap on the first word after "Throws:"? It seems like the word should have an initial cap. However, I don't remember what we've done previously and I don't want to cause a lot of extra work.

Author: I looked, we don't really have previous cases that are similar.

Creates a new CookieAuthenticationCredential instance. This function is used when importing sas-basic-auth globally.

See: [CDN](#CDN)

- `configuration: object`
  - `url: string` The URL of SAS Viya server you are authenticating against.
  - `guest: boolean` Automatically log in as a guest, if no user session is found.
    - default value: `false

## CookieAuthenticationCredential

The CookieAuthenticationCredential class provides functions to check the authentication status of a given endpoint

### `new CookieAuthenticationCredential(configuration)`

- `configuration: object`
  - `url: string` The URL of SAS Viya server you are authenticating against.
  - `guest: boolean` Automatically log in as a guest, if no user session is found.
    - default value: `false`

#### Methods

##### ` (): Promise<void>`

Checks to see if any user is authenticated.
Throws: if no user is authenticated.

##### `loginPopup(): Promise<void>`

Launches a popup window that navigates to the authentication endpoint and allows the user to login.
rejects: When login fails (for example: user closes the popup).

##### `invalidateCache()`

> If `checkAuthenticated` has already been called, a cached value may be returned the next time . This is done to reduce the number of potential service calls needed when making multiple API calls. Calling `invalidateCache` will force `checkAuthenticated` to re-validate the next time it is called. This function is not needed in most use cases. **This is not need for most use cases.**

Reviewer 3: Extra space after "time"

> If `checkAuthenticated` has already been called, a cached value may be returned the next time . This is done to reduce the number of potential service calls needed when making multiple API calls. Calling `invalidateCache` will force `checkAuthenticated` to re-validate the next time it is called. This function is not needed in most use cases. **This is not need for most use cases.**

Reviewer 3: Last sentence is repeated.

# API

## `sasAuthBrowser.createCookieAuthenticationCredential(configuration): CookieAuthenticationCredential`

Creates a new CookieAuthenticationCredential instance. This function is used when importing sas-basic-auth globally.

See: [CDN](#CDN)

- `configuration: object`
  - `url: string` The url of SAS Viya server you are authenticating against.
  - `guest: boolean` Automatically log in as a guest, if no user session is found.
    - default value: `false
## CookieAuthenticationCredential

The CookieAuthenticationCredential class provides functions to check the authentication status of a given endpoint

### `new CookieAuthenticationCredential(configuration)`

- `configuration: object`

>   - `url: string` The url of SAS Viya server you are authenticating against.

Reviewer 2: URL needs to be in all caps.

##### `invalidateCache()`

> If `checkAuthenticated` has already been called, a cached value may be returned the next time. This is done to reduce the number of potential service calls needed when making multiple API calls. Calling `invalidateCache` will force `checkAuthenticated` to re-validate the next time it is called. **This is not need for most use cases.**

Reviewer 2: In the last sentence, change "need" to "needed".
